metadata_service/proxy/atlas_proxy.py

```diff
@@ -476,6 +479,95 @@ def get_table(self, *, table_uri: str) -> Table:
                              'are missing in : ( {table_uri} )'
                              .format(table_uri=table_uri))
 
+    @staticmethod
+    def _validate_date(text_date: str, date_format: str) -> Tuple[Optional[datetime.datetime], Optional[str]]:
+        try:
+            return datetime.datetime.strptime(text_date, date_format), date_format
+        except (ValueError, TypeError):
+            return None, None
+
+    @staticmethod
+    def _select_watermark_format(partition_names: List[str]) -> Optional[str]:
+        result = None
+
+        for partition_name in partition_names:
+            # Assume that all partitions for given table have the same date format. Only thing that needs to be done
+            # is establishing which format out of the supported ones it is and then we validate every partition
+            # against it.
+            for df in app.config['WATERMARK_DATE_FORMATS']:
+                _, result = AtlasProxy._validate_date(partition_name, df)
+
+                if result:
+                    LOGGER.debug('Established date format', extra=dict(date_format=result))
+                    return result
+
+        return result
+
+    @staticmethod
+    def _render_partition_key_name(entity: EntityUniqueAttribute) -> Optional[str]:
+        _partition_keys = []
+
+        for partition_key in entity.get('attributes', dict()).get('partitionKeys', []):
+            partition_key_column_name = partition_key.get('displayName')
+
+            if partition_key_column_name:
+                _partition_keys.append(partition_key_column_name)
+
+        partition_key = ' '.join(_partition_keys).strip()
+
+        return partition_key
+
+    def _get_table_watermarks(self, entity: EntityUniqueAttribute) -> List[Watermark]:
+        partition_value_format = '%Y-%m-%d %H:%M:%S'
+
+        _partitions = entity.get('relationshipAttributes', dict()).get('partitions', list())
+        partition_key = AtlasProxy._render_partition_key_name(entity)
+
+        guids = [_partition.get('guid') for _partition in _partitions
+                 if _partition.get('entityStatus') == Status.ACTIVE
+                 and _partition.get('relationshipStatus') == Status.ACTIVE]
+
+        if not guids:
+            return []
+
+        full_partitions = extract_entities(self._driver.entity_bulk(guid=list(guids), ignoreRelationships=True))
+        watermark_date_format = AtlasProxy._select_watermark_format([p.attributes.get('name') for p in full_partitions])
```

Comment:
Do you really need to do this functionality here?
Since you are already validating the data on line 543, i.e., 
`partition_date, _ = AtlasProxy._validate_date(partition_name, watermark_date_format)` 
and on line 545 already making a check if partition_date, you can also make the check for watermark_date_format there if you want. 

This seems like an extra effort and extra loop as well on the full_partition

Reply:
I think that this is actually more efficient. The assumption is that we support any number different `WATERMARK_DATE_FORMATS` but a particular table should have consistent partition naming convention.

So table A might have partitions aligning to %Y%m%d format and tables B partitions would adhere to %Y,%m format. And if table A has some partitions aligning to %Y%m%d and some to %Y-%m-%d then we should ignore the latter ones.

After making this assumption - establishing this watermark_date_format in 534 acutally means that in 543 we don't iterate through all supported date formats for each partition name, just check if the name adheres to the particular one.